xl-objects-multipart.go

```diff
@@ -127,22 +141,44 @@ func (xl xlObjects) CompleteMultipartUpload(bucket string, object string, upload
 			Size:       fi.Size,
 		})
 		metadata.Size += fi.Size
+	}
+
+	// Loop through and atomically rename the parts to their actual location.
+	for index, part := range parts {
+		wg.Add(1)
+		go func(index int, part completePart) {
+			defer wg.Done()
+			partSuffix := fmt.Sprintf("%.5d.%s", part.PartNumber, part.ETag)
+			multipartPartFile := path.Join(mpartMetaPrefix, bucket, object, uploadID, partSuffix)
+			tmpMultipartObjSuffix := path.Join(tmpMetaPrefix, bucket, object, partNumToPartFileName(part.PartNumber))
+			rErr := xl.storage.RenameFile(minioMetaBucket, multipartPartFile, minioMetaBucket, tmpMultipartObjSuffix)
+			if rErr != nil {
+				errs[index] = rErr
+				return
+			}
+			multipartObjSuffix := path.Join(object, partNumToPartFileName(part.PartNumber))
+			rErr = xl.storage.RenameFile(minioMetaBucket, tmpMultipartObjSuffix, bucket, multipartObjSuffix)
+			if rErr != nil {
+				if dErr := xl.storage.DeleteFile(minioMetaBucket, tmpMultipartObjSuffix); dErr != nil {
+					errs[index] = dErr
+					return
```

Comment:
we should log each error here. If we have multiple errors, we will still notice only the first error.